transport/socket_transport.go

```diff
@@ -0,0 +1,503 @@
+		}
+		if !ignoreErr {
+			log.Errorf("error in reading from server connection: %v", err)
+		}
+		if err := c.conn.Close(); err != nil {
+			// Ignore
+		}
+	}
+	c.cleanUp()
+}
+
+func (c *serverConnection) cleanUp() {
+	c.s.removeConnection(c)
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.closed = true
+}
+
+func (c *serverConnection) readPanicHandler() {
+	// We use a custom panic handler as we don't want the server to panic and crash if it receives a malformed
+	// request which has insufficient bytes in the buffer which would cause a runtime error: index out of range panic
+	if r := recover(); r != nil {
+		log.Errorf("failure in connection readLoop: %v", r)
+		if err := c.conn.Close(); err != nil {
+			// Ignore
+		}
+		c.cleanUp()
+	}
+}
+
+func readMessage(conn net.Conn, messageHandler func([]byte) error) error {
+	buff := make([]byte, readBuffSize)
+	var err error
+	var readPos, n int
+	for {
+		// read the message size
+		bytesRequired := 4 - readPos
+		if bytesRequired > 0 {
+			n, err = io.ReadAtLeast(conn, buff[readPos:], bytesRequired)
+			if err != nil {
+				break
+			}
+			readPos += n
+		}
+		totSize := 4 + int(binary.BigEndian.Uint32(buff))
+		bytesRequired = totSize - readPos
+		if bytesRequired > 0 {
+			// If we haven't already read enough bytes, read the entire message body
+			if totSize > len(buff) {
+				// buffer isn't big enough, resize it
+				nb := make([]byte, totSize)
+				copy(nb, buff)
+				buff = nb
+			}
+			n, err = io.ReadAtLeast(conn, buff[readPos:], bytesRequired)
+			if err != nil {
+				break
+			}
+			readPos += n
+		}
+		// Note that the buffer is reused so it's up to the protocol structs to copy any data in the message such
+		// as records, uuid, []byte before the call to handleMessage returns
+		err = messageHandler(buff[4:totSize])
+		if err != nil {
+			break
+		}
+		remainingBytes := readPos - totSize
+		if remainingBytes > 0 {
+			// Bytes for another message(s) have already been read, don't throw these away
+			if remainingBytes < totSize {
+				// we can copy directly as no overlap
+				copy(buff, buff[totSize:readPos])
+			} else {
+				// too many bytes remaining, we have to create a new buffer
+				nb := make([]byte, len(buff))
+				copy(nb, buff[totSize:readPos])
+				buff = nb
+			}
+		}
+		readPos = remainingBytes
+	}
+	if err == io.EOF {
+		return nil
+	}
+	return err
+}
+
+func (c *serverConnection) handleMessage(buff []byte) error {
+	version := SocketTransportVersion(binary.BigEndian.Uint16(buff))
+	if version != SocketTransportV1 {
+		return errors.Errorf("invalid transport version: %d - only version %d supported", version, SocketTransportV1)
+	}
+	correlationID := binary.BigEndian.Uint64(buff[2:])
+	handlerID := int(binary.BigEndian.Uint64(buff[10:]))
+	handler, ok := c.s.getRequestHandler(handlerID)
+	if !ok {
+		return errors.Errorf("no handler found with id %d", handlerID)
+	}
+	responseBuff := make([]byte, 15, responseBuffInitialSize)
```

Comment:
reading this again, i am beginning to more strongly think the docs we talked about in a previous PR which documents the high-level byte structure goes a long way. 

I think the fact it goes out of date is less concerning. Its main value is giving some definition to every field in the buffer - like e.g correlationID - I am not sure what it is from reading this quickly.

Was the out-of-date thing the only pushback you had to documenting these things?

Reply:
> e.g correlationID - I am not sure what it is from reading this quickly.

It has the same meaning as the CorrelationID in the Kafka protocol. When a request is sent the client gives it a correlation id. When the server sends back a response for a particular request it includes the correlation id. This allows the client to match the response with the request - responses can come back in a different order to the order in which requests were sent.

Reply:
I will add some source code docs explaining the layout :)